### PR TITLE
Bring scream-homme interface back to life.

### PR DESCRIPTION
**Pull request by bartgol**
This PR brings Homme (including theta xx) back to be fully supported in scream. Some work was needed in the components/homme folder in order to make this possible; mostly, this adds up to some cmake changes in homme, and a few f90 routines being broken down into smaller pieces in Homme. A PR with all the homme changes has also been submitted in the e3sm repo. I will take care of handling conflicts with the ensuing upstream merge. We decided to go ahead and issue this PR to keep development going in SCREAM.

### CODE REVIEW

**Comment by bartgol** [comment]
Turns out, in order for test-all-scream to work correctly, this PR needs to wait for #727 to go in.

**Comment by bartgol** [comment]
@jgfouca I pinged you given the couple of mods in test-all stuff, in case you wanted to take a look. Feel free to ignore.

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp [hunk 1]
```diff
@@ -78,23 +100,24 @@ void HommeDynamics::set_grids (const std::shared_ptr<const GridsManager> grids_m
...
-  m_computed_fields.emplace("dp", dyn_scalar_3d_mid_layout, Pa,  dyn_grid_name);
-  m_computed_fields.emplace("qdp",dyn_tracers_layout,       Qdp, dyn_grid_name);
-
-  const int ftype = get_homme_param_value<int>("ftype");
+  m_computed_fields.emplace("u",       dyn_vector_3d_mid_layout, m/s, dyn_grid_name);
+  m_computed_fields.emplace("vtheta",  dyn_scalar_3d_mid_layout, K,   dyn_grid_name);
+  m_computed_fields.emplace("phi",     dyn_scalar_3d_int_layout, Pa*pow(m,3)/kg, dyn_grid_name);
+  m_computed_fields.emplace("w",       dyn_scalar_3d_int_layout, m/s, dyn_grid_name);
+  m_computed_fields.emplace("dp",      dyn_scalar_3d_mid_layout, Pa,  dyn_grid_name);
+  m_computed_fields.emplace("qdp",     dyn_tracers_layout,       Qdp, dyn_grid_name);
```
Comment (AaronDonahue) [inline components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:111]: I'm guessing that we plan to add the physics tendencies as a field later when we hook up physics + dynamics?

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp [hunk 2]
```diff
@@ -16,19 +11,46 @@
...
-  init_homme1(comm);
+  // Init homme context par struct
+  if (!is_parallel_inited_f90()) {
+    auto comm_f = MPI_Comm_c2f(comm.mpi_comm());
+    init_parallel_f90(comm_f);
+  }
+
+  // Load homme params from namelist
+  if (!is_params_inited_f90()) {
+    auto nl_fname = m_params.get<std::string>("Homme namelist filename","namelist.nl");
```
Comment (AaronDonahue) [inline components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:38]: Should the default namelist be "user_nl_cam"?

File: components/scream/src/dynamics/homme/homme_inputs_initializer.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+void HommeInputsInitializer::initialize_fields () {
+
+  // We cheat a little bit here. It is way easier for Homme to init
+  // fields during the whole prim initialization sequence.
+  // However, such sequence happens during calls to HommeDynamics'
+  // ctor and initialize method. The AD attempts to init the
+  // inputs *after* all atm process have been inited.
+  // So here's what we do: if the user asks to init inputs,
+  // we simply go ahead and to it, by calling this method inside
```
Comment (AaronDonahue) [inline components/scream/src/dynamics/homme/homme_inputs_initializer.cpp:22]: "go ahead and **do** it"

File: components/scream/src/dynamics/homme/homme_inputs_initializer.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  // HommeDynamics' initialize call. When later the AD calls this
+  // initializer, we simply skip everything, since we already inited
+  // all the fields.
+  // Note: if some other atm proc is trying to init any of the fields
+  //       inited by this class, we the AD should throw an error anyways.
```
Comment (AaronDonahue) [inline components/scream/src/dynamics/homme/homme_inputs_initializer.cpp:27]: **we** or **the AD** ...

File: components/scream/src/dynamics/homme/homme_inputs_initializer.cpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  if (!m_fields_inited) {
+    // Sanity check. This class should init *all* homme's inputs:
```
Comment (AaronDonahue) [inline components/scream/src/dynamics/homme/homme_inputs_initializer.cpp:30]: Is there a reason all of this commented code is left in?  Do we plan to revitalize it later on?

File: components/scream/tests/scream_homme_dyn_p3/CMakeLists.txt [hunk 6]
```diff
@@ -2,36 +2,34 @@ INCLUDE (ScreamUtils)
 
 # Get or create the dynamics lib
 #                 HOMME_TARGET   NP PLEV QSIZE_D USE_PIO USE_ENERGY 
-CreateDynamicsLib("preqx_kokkos"  4   72   35    FALSE   FALSE)
+CreateDynamicsLib("theta-l_kokkos"  4   72   35    FALSE   FALSE)
```
Comment (AaronDonahue) [inline components/scream/tests/scream_homme_dyn_p3/CMakeLists.txt:5]: Here you keep the USE_PIO and USE_ENERGY columns, but not elsewhere.  Is there a reason why?

File: components/scream/tests/scream_homme_dyn_ut_nlev72_qsize4/CMakeLists.txt [hunk 7]
```diff
@@ -1,44 +1,56 @@
-INCLUDE (ScreamUtils)
+include (ScreamUtils)
 
 # Get or create the dynamics lib
-#                 HOMME_TARGET   NP PLEV QSIZE_D USE_PIO USE_ENERGY 
-CreateDynamicsLib("preqx_kokkos"  4   72   35    FALSE   FALSE)
+#                 HOMME_TARGET   NP PLEV QSIZE_D
```
Comment (AaronDonahue) [inline components/scream/tests/scream_homme_dyn_ut_nlev72_qsize4/CMakeLists.txt:4]: For example her, no USE_PIO and USE_ENERGY

**Comment by AaronDonahue** [comment]
Final comment.  Every time I've seen this PR title, the song "Wake me up" by Evanescence gets stuck in my head.

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp (hunk 1 above)
Comment (bartgol) [reply components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:111]: Yes, I haven't really put much effort into adding all the fields that Dynamics will need. My primary goal was to have homme compiled in, and working. AD hooks can be adjusted later.

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp (hunk 2 above)
Comment (bartgol) [reply components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:38]: I used the name that homme uses as default in its internal tests. If you prefer, we could even remove the default (causing a runtime error if the nl name is not set in the input yaml file...

File: components/scream/src/dynamics/homme/homme_inputs_initializer.cpp (hunk 5 above)
Comment (bartgol) [reply components/scream/src/dynamics/homme/homme_inputs_initializer.cpp:30]: Ah, good catch. The answer is 'maybe?'... I can remove it for now, and forget about checks like this one. After all, the names for the fields are also going to change (to embrace CF ones), so might as well do everything later. I'll add a TODO comment as a reminder.

File: components/scream/tests/scream_homme_dyn_p3/CMakeLists.txt (hunk 6 above)
Comment (bartgol) [reply components/scream/tests/scream_homme_dyn_p3/CMakeLists.txt:5]: Ah, this test is not even compiled, so I didn't notice the old signature of the macro. I will fix this, but keep in mind this folder is skipped anyways.

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp (hunk 2 above)
Comment (AaronDonahue) [reply components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:38]: I would vote for deleting the default and throwing an error.  A physics developer won't know about the homme internal tests and may get the impression that a namelist.nl file is required.  If we decide to include a default namelist with the codebase we can always add the default name back in here.  Thoughts?

File: components/scream/src/dynamics/homme/atmosphere_dynamics.cpp (hunk 1 above)
Comment (AaronDonahue) [reply components/scream/src/dynamics/homme/atmosphere_dynamics.cpp:111]: Okay cool.  That is what I figured, but wanted to make sure I mentioned it.  We may want to create a git issue as well so we can keep track of the big ToDo items.

**Review by AaronDonahue — APPROVED** [review]
Looks good, and a lot of work here.  Really looking forward to getting this into the main code and taking it for a few test drives.  Thanks Luca!

**Comment by bartgol** [comment]
After a long fight with slurm and cmake, I think I finally managed to address the last outstanding thechnical issues (ping me if you want details).

I'll wait for the AT to confirm everything is working, then I'll ping you guys again.

**Review by jgfouca — APPROVED** [review]

**Pull request merged**